[FLINK-13905][checkpointing] Separate checkpoint triggering into several asynchronous stages

ifndef-SleePy opened the pull request:
## What is the purpose of the change

* This PR aims to separating checkpoint triggering into several asynchronous stages.
* This is the heart part of threading model refactoring.

## Brief change log

* Unify all error handling of checkpoint failure.
* Change the semantic from queued to suspended if there are too many in-flight checkpoints.
* Separate the triggering from synchronous to asynchronous.
  * The last commit is a bit large although I have done some preparations in the prior and this PR :(
  * Introduces an `isTriggering` variable to avoid competition between in-flight checkpoint trigger and the later one. Because we have to keep the order of triggering and avoid concurrent competition.
  * Introduces a `triggerRequestQueue` which is used to queue the trigger request which can't be triggered at that moment.
  * There are only two exits of triggering, `onTriggerSuccess` and `OnTriggerFailure`. Either of them must be called at the end of triggering lifecycle. In there, we maintain the status of `isTriggering`.
  * Do not invoke `CheckpointFailureManager#handleJobLevelCheckpointException` when trigger failure. It's useless for now. And it's also a blocking operation to get the current checkpoint id.
  * There are lots of test cases need to change because `triggerCheckpoint` becomes an asynchronous operation. We need to manually trigger the asynchronous trigger stages.
  * `triggerCheckpoint` will not throw exception directly even it fails at the very beginning. All error handling outside checkpoint coordinator should be through return value of `triggerCheckpoint`. In this way, we don't need "try catch" everywhere.

## Verifying this change

* There are a lot of existing test cases and this change also added tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: it should be no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

ifndef-SleePy commented:
@flinkbot attention @pnowojski

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -708,22 +689,33 @@ public void receiveDeclineMessage(DeclineCheckpoint message, String taskManagerL
+					checkpointException =
+						new CheckpointException(
+							CheckpointFailureReason.JOB_FAILURE, message.getReason());
+				}
```
extract to a helper method?

ifndef-SleePy replied:
OK

pnowojski replied:
Has this happened? I can not find this fixed in the the last `[hotfix] Address review comments.` nor it's fixed in this commit.

Btw, having one commit with all of the addressed comments, where comments were referring to multiple commits is not the best solution. Since you would have to fixup the correct commits before merging, it would be better to have couple of smaller fixup commits (one fixup commit per original commit). (Also FYI, I would be fine with you already squashing the fixup commits with the original commits)

ifndef-SleePy replied:
Ah, I forgot to leave an explanation about this. I didn't squash the original commits to make it easy to continue reviewing.

Actually I have extracted the "else if ... else" to a helper method named `getCheckpointException`. But it does happen in fixup commit but in the original commit "[FLINK-13905][checkpointing] Separate checkpoint triggering into several asynchronous stages.". I thought it's enough as a helper method.

And thanks for the reminding to keep per fixup as a separate commit. I'll try to separate the last big fixup commit :)

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1262,11 +1261,19 @@ public void stopCheckpointScheduler() {
-			for (PendingCheckpoint p : pendingCheckpoints.values()) {
-				failPendingCheckpoint(p, exception.getCheckpointFailureReason());
-			}
+			// do not traverse pendingCheckpoints directly, because it might be changed
```
? Is this a bug fix?

ifndef-SleePy replied:
No, it's a reminder. It's introduced by this PR.
After the refactoring of error handling, the `failPendingCheckpoint` is the only one entrance to fail a pending checkpoint. The `failPendingCheckpoint` maintains the `pendingCheckpoints`. We don't need to maintain `pendingCheckpoints` everywhere. However as side-effect, we can't invoke `failPendingCheckpoint` in a for loop like before. A `ConcurrentModificationException` will be raised in that way.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -957,17 +953,19 @@ private void rememberRecentCheckpointId(long id) {
+
+		abortPendingCheckpoints(
+			checkpointsToSubsume,
+			new CheckpointException(CheckpointFailureReason.CHECKPOINT_SUBSUMED));
```
For one thing, you could pass a stream as a parameter to the `abortPendingCheckpoints`, to deduplicate code a bit. However to be honest I do not see much value of keeping `abortPendingCheckpoints(PendingCheckpoint[] checkpoints, CheckpointException exception)` and I think I would just inline it.

Having it, deduplicates only a simple `for (PendingCheckpoint pendingCheckpoint : checkpoints) ` loop, while callers still have to iterate over the checkpoints one way or another (creating a stream, filtering it, collecting to an array).

ifndef-SleePy replied:
The purpose of passing an array to `abortPendingCheckpoints` as parameter is to avoid potential `ConcurrentModificationException`. There are more details in the response above.

ifndef-SleePy replied:
Do you mean providing a `abortPendingCheckpoint(Predicate<PendingCheckpint> checkpointToDeletePredicate)`,  instead of calling `pendingCheckpoints.remove(pendingCheckpoint.getCheckpointId())` directly in `abortPendingCheckpoint`?

I guess it works. However I'm not sure we should do that or not. The codes might be hard to understand. It seems that we have to make a decision about how to trade off this.

ifndef-SleePy replied:
I think I got your point now :)
Makes sense to me.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1262,11 +1261,19 @@ public void stopCheckpointScheduler() {
+	}
+
+	private void abortPendingCheckpoints(
+		PendingCheckpoint[] checkpoints, CheckpointException exception) {
```
nit: if one parameter is indented, all of them should be (according to our code style)

ifndef-SleePy replied:
Alright, I need to re-check the code style guide :)

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -648,24 +636,17 @@ public boolean isShutdown() {
-			}
-
-			try {
-				checkpointStorageLocation.disposeOnFailure();
```
What has happened to this line? Was it replaced by something?

ifndef-SleePy replied:
It's duplicated. `abortPendingCheckpoint` -> `PendingCheckpoint#dispose` will invoke `CheckpointStorageLocation#disposeOnFailure`. That's the benefit of unifying error handling into `abortPendingCheckpoint`. We don't have to release resource everywhere.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -363,10 +361,9 @@ public void shutdown(JobStatus jobStatus) throws Exception {
-				pendingCheckpoints.clear();
+				abortPendingCheckpoints(
+					new CheckpointException(
+						CheckpointFailureReason.CHECKPOINT_COORDINATOR_SHUTDOWN));
```
it looks like before `triggerQueuedRequests` was not called here (and in some other places), while it's called now?

ifndef-SleePy replied:
Yes, but `checkAndResetCheckpointScheduler` is called before. They are both trying to resume periodic triggering.
Anyway, `triggerQueuedRequests` will not work here even it's called. Because it will check `shutdown` and `periodicScheduling` first.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1352,52 +1359,6 @@ public void run() {
-
-		if (!haveMoreRecentPending) {
-			triggerQueuedRequests();
-		}
```
This if check is gone now? Now this is called always?

ifndef-SleePy replied:
Please see the response below. I changed the behavior a bit of resuming periodic triggering.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1429,61 +1390,45 @@ public void run() {
-			synchronized (lock) {
-				if (pendingCheckpoints.isEmpty() || allPendingCheckpointsDiscarded()) {
-					triggerRequestQueued = false;
-					currentPeriodicTrigger = scheduleTriggerWithDelay(getRandomInitDelay());
```
what has happened to this logic?

ifndef-SleePy replied:
Basically `checkAndResetCheckpointScheduler` does same thing with `triggerQueuedRequests` (try to resume the periodic triggering) but under a stricter condition (there is no more in-flight checkpoint). 
`checkAndResetCheckpointScheduler` is invoked when an in-flight checkpoint fails. `triggerRequestQueued` is invoked when an in-flight checkpoint succeeds or timeouts.

It does not makes much sense but it makes the codes more complicated. I think the rule here could be simplified a lot.

It suspends the periodic triggering when there are too many in-flight checkpoints. So it should resume periodic triggering when there are fewer in-flight checkpoints than `maxConcurrentCheckpointAttempts`. That means trying to resume periodic triggering when an in-flight checkpoint ends its lifecycle, no matter succeeding or failing.

pnowojski replied:
ok, it makes sense to me.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -162,9 +162,9 @@
-	private boolean triggerRequestQueued;
+	/** Flag whether periodic triggering is suspended (too many concurrent pending checkpoint).
+	 * Non-volatile, because only accessed in synchronized scope */
+	private boolean periodicTriggeringSuspended;
```
Could you rephrase and elaborate the commit message
```
Actually the newcome trigger request will be failed if there are too many pending checkpoints.
It's more like suspending periodic triggering rather than queuing.
And the semantic of queuing will be used in later commit.
```
? I'm struggling to understand the motivation behind this commit?

ifndef-SleePy replied:
OK let me explain more about the background.
1. I want to introduce a queue to cache those trigger requests that can't be triggered for now. Because there is an asynchronous in-flight trigger. This is described in last commit.
2. I found there is already a queue logic existing, `triggerRequestQueued`.
3. I tried to unify these two queuing concepts. But I found it makes things much more complicated. It's hard to understand or maintain the queuing logic without this background. I believe it's more reasonable to keep both these queuing concepts separately.
4. I renamed the `triggerRequestQueued` to `periodicTriggeringSuspended` to distinguish these two concepts. I think the `triggerRequestQueued` relevant concept is not correct somewhat. It does not really queue the trigger request but fail it and suspend the periodic triggering instead.

pnowojski replied:
Ok, thanks for the extra context (this explanation I think also could be added to this commit message). I was also confused/didn't like the previous `triggerRequestQueued` name.

However I'm still missing if this commit changes any behaviour or is it just pure rename/clean up?

ifndef-SleePy replied:
I have explained a bit more in the response inline. It's not just a pure renaming. To be honest,  after so many changes the codes are not easy to maintain. So I did a bit refactoring when I believe it's necessary.

ifndef-SleePy replied:
Good suggestion. Will do that later.

ifndef-SleePy commented:
Hi @pnowojski , thanks for reviewing! Please check my response inline.

BTW, I didn't update the PR to address comments yet. Because I think it's better to do it till all commits are reviewed.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -969,25 +969,25 @@ private void dropSubsumedCheckpoints(long checkpointId) {
-				currentPeriodicTrigger = scheduleTriggerWithDelay(0L);
-			}
-			else {
-				timer.execute(new ScheduledTrigger());
```
Why has this behaviour disappeared?

ifndef-SleePy replied:
Actually there is a difference between `triggerQueuedRequests` and `resumePeriodicTriggering`. It's about the non-periodic checkpoint triggering (not savepoint). However there is no such a scenario (non-periodic checkpoint) in production codes. It's dead code somewhat.

This part is about the non-periodic checkpoint. In the `triggerQueuedRequests` version, it would re-schedule a one-shot checkpoint trigger if `periodicScheduling` is false. I guess the reason is that it assume the `triggerRequestQueued` is triggered by a non-periodic checkpoint trigger. So it recovers a non-periodic trigger. While in the `resumePeriodicTriggering` version, it would do nothing if periodic triggering is disabled.

Let's assume there is a non-periodic trigger scenario. Re-scheduling an one-shot trigger also makes no sense to me. I don't think this behavior worth being kept. If caller wants an one-shot checkpoint being executed anyway, the `forceCheckpoint` should be set. Otherwise this one-shot checkpoint might fail and will never be re-scheduled. Caller could track the lifecycle thought `CompletableFuture` returned by `triggerCheckpoint`.

ifndef-SleePy replied:
I guess your concern is that some necessary trigger might be missing?
Stopping with savepoint would cancel all in-flight checkpoints/savepoints. And it does not allow any further periodic triggering. The `ScheduledTrigger` is just a normal one-shot checkpoint. I think it's fine to remove it. I didn't find any reason to keep it. I believe the reason it stays here is the unclear semantics of `triggerQueuedRequests`. There is a trigger request "queued", we should "dequeue" something afterwards.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1441,13 +1441,7 @@ private void preCheckBeforeTriggeringCheckpoint(boolean isPeriodic, boolean forc
-				LOG.warn("Trying to trigger another checkpoint for job {} while one was queued already.", job);
-				throw new CheckpointException(CheckpointFailureReason.ALREADY_QUEUED);
-			}
-
```
And why is this no longer possible?

ifndef-SleePy replied:
It's possible. But since I have refactored the `triggerRequestQueued` to `periodicTriggeringSuspended`, it makes no more sense to check `periodicTriggeringSuspended`. If the trigger could satisfy other requirements, it could be executed. 

Actually in the current implementation. It's also about the non-periodic checkpoint triggering.
1. Savepoint does not care this checking, because  the `forceCheckpoint` is set to true.
2. Periodic checkpoint trigger would not reach here because it would be canceled if `triggerRequestQueued` is set to true. (There should be no in-flight periodic trigger since it's single-threaded.)
So it's also dead code...

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -704,12 +882,9 @@ public void receiveDeclineMessage(DeclineCheckpoint message, String taskManagerL
 					checkpointException =
 						new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED);
-				} else if (message.getReason() instanceof CheckpointException) {
-					checkpointException = (CheckpointException) message.getReason();
```
I guess this is no longer needed and that's why it was removed?

ifndef-SleePy replied:
I unified this with the "else" below through `getCheckpointException`.

pnowojski replied:
Ops, right. Sorry I missed that :)

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1248,8 +1427,10 @@ public void stopCheckpointScheduler() {
+			final CheckpointException reason =
+				new CheckpointException(CheckpointFailureReason.CHECKPOINT_COORDINATOR_SUSPEND);
+			abortQueuedTriggerRequests(reason);
+			abortPendingCheckpoints(reason);
```
nit: maybe extract those lines to something like
```
abortPendingAndQueuedCheckpoints(CheckpointFailureReason.CHECKPOINT_COORDINATOR_SUSPEND);
```

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -484,77 +484,158 @@ public boolean isShutdown() {
+					LOG.info("Triggering checkpoint {} @ {} for job {}.",
+						checkpointIdAndStorageLocation.checkpointId, timestamp, job);
+					return snapshotMasterState(pendingCheckpoint);
+				}, timer)
```
> The end goal for is to replace timer executor with the JobManager's main thread executor, right?

Yes, you are right.

> What's blocking us from using main thread executor vs timer right now?

The blocker is the coordinator-wide lock. There are still some heavy operations under this lock. For example, handling ACK/declined message in IO threads. We shouldn't make main thread waiting for this heavy lock, but timer thread is fine. Although currently there are some scenarios that it waits this lock in main thread executor (shutting down checkpoint coordinator or starting/stopping checkpoint scheduler). These operations (shutdown/starting/stopping) do not happen frequently, so there is no big trouble yet. But it's hard to say what will happen if we put more frequent trigger logic into main thread now.

After this and next PR (refactoring threading model of handling ACK/declined message), we could avoid the coordinator-wide lock then (the coordinator-wide lock would not be used in IO thread anymore). That's the time to replace timer thread with main thread.

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -575,86 +657,182 @@ public boolean isShutdown() {
+				return;
+			}
+		}
+		timer.execute(() -> {
```
I guess you are right. It seems that all callers are in timer thread.

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -484,77 +484,158 @@ public boolean isShutdown() {
+			// keep a reference outside the async calls
+			final AtomicReference<PendingCheckpoint> currentCheckpointRef = new AtomicReference<>();
+			initializeCheckpoint(props, externalSavepointLocation)
+				.thenComposeAsync((checkpointIdAndStorageLocation) -> {
```
Sounds reasonable. I could make it shorter.

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -484,77 +484,158 @@ public boolean isShutdown() {
+							onTriggerFailure(currentCheckpointRef.get(), throwable);
+						}
+					}
+				}, timer);
```
> Why are you splitting this into a three chain calls after initializeCheckpoint? Can not this be initializeCheckpoint -> "everything else at once" ?

Because snapshotting master hooks is also asynchronous. So there is a `thenComposeAsync` + `thenRun`. The last `whenCompleteAsync` is for error handling of all these async chains. Actually `exceptionally` will be easier to read. However the error handling must be executed in timer thread, there is no such an `exceptionallyAsync` available.

> Also final AtomicReference<PendingCheckpoint> currentCheckpointRef = new AtomicReference<>(); seems strange having to use AtomicReference for something that's basically a synchronous chain of calls one after another, right?

The `AtomicReference` is used to refer the `PendingCheckpoint` in lambda expressions (chain calls and error handling). It has to be a final variable. Let me try to make it more readable.

pnowojski replied:
So it's the following chain of actions:

1. `initializeCheckpoint` in the `ioExecutor`
2. `createPendingCheckpoint` in the `timer`
3. `triggerHook` in the `ioExecutor`
4. follow up of `triggerHook` (lines 694:717 from CheckpointCoordinator) in `timer`
5.`snapshotTaskState` and `onTriggerFailure` in the `timer`?

If we squashed 4 and 5, wouldn't that simplify the code?

ifndef-SleePy replied:
The part 4 is quite independent. And I think someday the part 4 could be put to another place, like being executed concurrently with snapshotting task state, or even being removed.
So I think we should make it as an independent method.

I reorganized this piece of codes a bit to make it shorter. I also removed the `AtomicReference` outside the chain calls.

ifndef-SleePy replied:
BTW, I replaced the `checkState(!checkpoint.isDiscarded());` by `!checkpoint.isDiscarded()`. Because I found there is a possibility that `PendingCheckpoint` might be cancelled during the asynchronous call if there is no master hook (otherwise `snapshotMasterState` would check state of `PendingCheckpoint`).

ifndef-SleePy commented:
Hi @pnowojski , 

> To put this into a context. Could you explain/refresh my memory what are you plans for next steps after this PR?

Next step is handling the ACK/declined message of `CheckpointCoordinator`, which is described in https://issues.apache.org/jira/browse/FLINK-14971. After that, we could get back to https://issues.apache.org/jira/browse/FLINK-13848. That's the last blocker of replacing timer thread with main thread executor. If all these go well, we could do the last work of this refactoring, removing coordinator-wide lock and replacing timer thread with main thread.

pnowojski commented:
Thanks for the explanations. Could you push the changes to address the issues/comments? I would do the second review pass after that.

ifndef-SleePy commented:
Hi @pnowojski , thanks for patiently reviewing. I have addressed most of the comments. Regarding to chain calls part, I tried to make it shorter and more readable. Could you please check it again?

pnowojski commented:
I'm checking it again. Posted just a single comment as I had to interrupt the reviewing.

ifndef-SleePy commented:
@pnowojski , got it. 

I have split the fixup commits into small ones.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -163,9 +163,9 @@
-	/** Flag whether a trigger request could not be handled immediately. Non-volatile, because only
-	 * accessed in synchronized scope */
-	private boolean triggerRequestQueued;
+	/** Flag whether periodic triggering is suspended (too many concurrent pending checkpoint).
```
nit: I think the grammar of the commit message is a bit confusing (what was in the past and will be in the future).

If I understand it correctly it should be:
```
This is not just a pure renaming. The semantics has been changed a bit.
If there are too many in-flight checkpoints, the current checkpoint
will fail, and the periodic triggering will be suspended. This stays the same 
as it used to be before.

From now on, there will be no queued semantics. If an in-flight checkpoint
ends its lifecycle, it will try to resume periodic triggering, but no
queued checkpoint will be resumed anymore, as it was in the past.
```
?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -485,76 +481,151 @@ public boolean isShutdown() {
-						ExecutionState.RUNNING,
-						ee.getState());
-				throw new CheckpointException(CheckpointFailureReason.NOT_ALL_REQUIRED_TASKS_RUNNING);
+			if (isTriggering || !triggerRequestQueue.isEmpty()) {
```
Just out of curiosity. Do we need to acquire the `lock` here? Isn't `isTriggering` and `triggerRequestQueue ` touched/used only in the `timer` thread?

ifndef-SleePy replied:
`triggerRequestQueue` might be accessed in main thread, through `shutdown` and `stopCheckpointScheduler`.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -485,76 +481,151 @@ public boolean isShutdown() {
+						isPeriodic,
+						checkpointIdAndStorageLocation.checkpointId,
+						checkpointIdAndStorageLocation.checkpointStorageLocation,
+						onCompletionPromise), timer);
```
nit: `timer` in new line? It's not easy to deduce where the `timer` parameter belongs to. Maybe:
```
	.thenApplyAsync(
		(checkpointIdAndStorageLocation) -> createPendingCheckpoint(
			timestamp,
			props,
			ackTasks,
			isPeriodic,
			checkpointIdAndStorageLocation.checkpointId,
			checkpointIdAndStorageLocation.checkpointStorageLocation,
			onCompletionPromise),
		timer);
```
and in other places (I think everywhere where you are specifying executor)?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -485,76 +481,151 @@ public boolean isShutdown() {
+							}
+						});
+					}
+				});
```
I'm a bit confused which thread is actually being used to execute what here. For example isn't the whole `whenComplete` executed already in the `timer` thread, since `CheckpointCoordinator#snapshotMasterState` returns a future that will be completed from the `timer` thread? 

If so, why do we need to call `timer.execute(()`?

Regardless if I'm wrong/right about what's executed in which thread, I think this my confusion shows something has to be changed here. Maybe always explicitly use the `executor`, `timer` parameter when scheduling/chaining something, so that the executor is clearly visible? `thenCompose(...)` -> `thenComposeAsync(..., timer/executor)`? I think that would be especially helpful as it's not easily visible/understandable what's returned from `this::snapshotMasterState`. Or at least add comments? But `selfDocumentingCode > comments`.

ifndef-SleePy replied:
> For example isn't the whole whenComplete executed already in the timer thread, since CheckpointCoordinator#snapshotMasterState returns a future that will be completed from the timer thread?

Yes, all these chain calls should be executed in timer thread, except the error handling part. The exceptional future might be completed in IO thread in `initializeCheckpoint`. 

> Maybe always explicitly use the executor, timer parameter when scheduling/chaining something, so that the executor is clearly visible? thenCompose(...) -> thenComposeAsync(..., timer/executor)? 

Yes, it is definitely clearer. Actually I had implemented it like this at the beginning. But it's not efficient for most scenarios, they are already in timer thread. Especially for the final implementation (replace timer thread by main thread), it's an unnecessary burden for main thread.

I think the perfect solution here is checking the current thread. If it's already in timer (or main) thread, it would be executed directly. Otherwise it would be scheduled in timer (or main) thread. However it's not easy (or a bit hacky) to do so for now, because timer is a `ScheduledExecutor`. My plan is to implement this after replacing the timer thread by main thread, through providing a method maybe named `ComponentMainThreadExecutor#isRunningInMainThread`.

pnowojski replied:
I'm not sure if performance matter here. This is after all a rare (at best one call every couple of seconds, more realistically one call every couple of minutes), with the overhead that should be in ~1/1000_000th of a second.

I would still go with the explicit passing of the executor, but if you prefer I would be fine with adding comments what is executed where, as this is really important for thinking about code correctness.

> Yes, all these chain calls should be executed in timer thread, except the error handling part. The exceptional future might be completed in IO thread in initializeCheckpoint.

If a failure originates from `initializeCheckpoint`, from `executor` (io), then `.whenComplete(...)` part (L551:574) are executed in the `executor` thread as well? In all of the other cases those same lines are executed in the `timer` thread? This is a bit confusing. Could we make it execute always in the same thread?

ifndef-SleePy replied:
Em... It's indeed a bit confusing. I think you have convinced me. Alright, let's make it simple, forget the tiny overhead.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -576,85 +648,181 @@ public boolean isShutdown() {
+	 * @param throwable the reason of trigger failure
+	 */
+	private void onTriggerFailure(@Nullable PendingCheckpoint checkpoint, Throwable throwable) {
+		try {
```
nit: maybe we could add some `checkState` that we are indeed in the `timer` thread? But I'm not sure if it's possible/easily do-able?

ifndef-SleePy replied:
Yes, you are right, we need some checking here. As mentioned above, it's not easy (or a bit hacky) to check current thread for a `ScheduledExecutor`. Since the timer thread would be abandoned when we put all these codes into main thread, I think we should not spending too much time on this. We could do that in main thread.

ifndef-SleePy commented:
@pnowojski , I have addressed the explicit comments. Regarding to the async calls in timer thread, I think we could leave it along here. Because that part is intermediate, if everything goes well, we could replace timer thread soon. What do you think?

pnowojski commented:
> @pnowojski , I have addressed the explicit comments. Regarding to the async calls in timer thread, I think we could leave it along here. Because that part is intermediate, if everything goes well, we could replace timer thread soon. What do you think?

But if we replace the `timer` thread with `mainThreadExecutor`, the code should stay more or less the same, just the executor will change here, or not?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -485,76 +481,152 @@ public boolean isShutdown() {
+					},
+					timer);
+		} catch (Throwable throwable) {
+			onTriggerFailure(onCompletionPromise, throwable);
```
Hmmm, are all of the invocations of `#startTriggeringCheckpoint(...)` coming from the `timer` thread? I think so, but If not, there would be a potential for race conditions between the calls to `onTriggerFailure()` (here) coming from that other thread and the timer thread.

ifndef-SleePy replied:
Yes, `startTriggeringCheckpoint` must be executed in timer thread. We can guarantee that now.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+			CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION,
+			true,
+			false,
+			0);
```
Could you deduplicate the code like this in this file? Either create a test (parameterless?) constructor of `CheckpointCoordinatorConfiguration`, or provide a `private static CheckpointCoordinatorConfiguration createCheckpointCoordinatorConfiguration() { ...}` method in this file, or even move `chkConfig` to `@Setup` or `@Before` method?

I think the same applies to `coord` variable.

ifndef-SleePy replied:
Yes, there are too many codes copied everywhere. I'll introduce some utils for these initializations.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+			true,
+			false,
+			0);
+		CheckpointCoordinator coord = new CheckpointCoordinator(
```
nit: please do not use abbreviations:
`chkConfig` -> `checkpointCoordinatorConfiguration`/`configuration`
`coord` -> `coordinator`/`checkpointCoordinator`

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
 	}
 
+	@Test
+	public void testTriggerCheckpointRequestQueued() throws Exception {
```
optional nit: rename `testTriggerCheckpointRequestQueued` -> `testTriggerCheckpointBeforePreviousOneCompleted`? (to name it after a behaviour/test case instead of a private field name?)

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		assertFalse(onCompletionPromise1.isCompletedExceptionally());
+		assertFalse(coord.isTriggering());
+		assertEquals(0, coord.getTriggerRequestQueue().size());
```
Wouldn't it be better to just manually trigger two checkpoints (instead of using periodic scheduling for the second one), and assert both completion futures? Since I think here we are just asserting the success of the manually scheduled checkpoint, not the periodic one?

ifndef-SleePy replied:
Good idea.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+		assertEquals(2, coord.getTriggerRequestQueue().size());
+
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		assertTrue(onCompletionPromise1.isCompletedExceptionally());
```
why has the first one failed?

ifndef-SleePy replied:
The `TestingCheckpointIDCounter` fails the checkpoint. I'll rename it and add a comment here.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+
+		coord.startCheckpointScheduler();
+		// Periodic
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise1 = coord.triggerCheckpoint(
```
nit: `onCompletionPromise1` -> `periodicCompletionPromise`?

ifndef-SleePy replied:
I think `onCompletionPromise1` is better here. It's not important whether it's periodic or not. The comment should be clearer.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		try {
+			onCompletionPromise1.get();
+			fail("Should not reach here");
```
What is causing a failure in this test case? I think this might be missing some comment in the lines above, what is causing the failure

ifndef-SleePy replied:
Hmmm, it's a bit hard to understand.
It's due to `TestingCheckpointIDCounter`. I should give it a better name.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+			CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
+			null,
+			true,
+			false);
```
maybe extract to a named method, like `onCompletionPromise = triggerPeriodicCheckpoint(coordinator)`? This way you would deduplicate some code and could remove the `// Periodic` comment (`selfDocumentingCode` > `comment`)

ifndef-SleePy commented:
@pnowojski , I have addressed review comments. Could you please check it again?

ifndef-SleePy commented:
@pnowojski , thanks for reviewing and nice suggestions. I have done the re-ordering. Put the commits
`Introduce builder for checkpoint coordinator ...`
`Normalize variable name of CheckpointCoordinatorTriggeringTest`
into the base of this PR. Regarding to the other two commits you mentioned, I found that they are basically affects the codes added in `[FLINK-13905][checkpointing] Separate checkpoint triggering into several asynchronous stages.`. So I squashed them into this last commit.
